src/algebra/commute.lean

```diff
@@ -0,0 +1,371 @@
+
+E.g., if `a`, `b`, and c are elements of a ring, and
+that `hb : commute a b` and `hc : commute a c`.  Then
+`hb.pow_left 5` proves `commute (a ^ 5) b` and
+`(hb.pow 2).add (hb.mul hc)` proves `commute a (b ^ 2 + b * c)`.
+
+Lean does not immediately recognise these terms as equations,
+so for rewriting we need syntax like `rw [(hb.pow_left 5).eq]`
+rather than just `rw [hb.pow_left 5]`.
+-/
+
+/-- Two elements commute, if `a * b = b * a`. -/
+def commute {S : Type*} [has_mul S] (a b : S) : Prop := a * b = b * a
+
+namespace commute
+
+variables {S : Type*} [has_mul S]
+
+/-- Equality behind `commute a b`; useful for rewriting. -/
+protected theorem eq {a b : S} (h : commute a b) : a * b = b * a := h
+
+/-- Any element commutes with itself. -/
+@[refl] protected theorem refl (a : S) : commute a a := eq.refl (a * a)
+
+/-- If `a` commutes with `b`, then `b` commutes with `a`. -/
+@[symm] protected theorem symm {a b : S} (h : commute a b) : commute b a :=
+eq.symm h
+
+protected theorem symm_iff {a b : S} : commute a b ↔ commute b a := ⟨commute.symm, commute.symm⟩
+
+end commute
+
+-- Definitions and trivial theorems about them
+section defs
+
+variables {S : Type*} [has_mul S]
+
+/-- Centralizer of an element `a : S` is the set of elements that commute with `a`. -/
+def centralizer (a : S) : set S := { x | commute a x }
+
+theorem mem_centralizer {a b : S} : b ∈ centralizer a ↔ commute a b := iff.rfl
+
+/-- Centralizer of a set `T` is the set of elements that commute with all `a ∈ T`. -/
+protected def set.centralizer (s : set S) : set S := { x | ∀ a ∈ s, commute a x }
+
+protected theorem set.mem_centralizer (s : set S) {x : S} :
+  x ∈ s.centralizer ↔ ∀ a ∈ s, commute a x :=
+iff.rfl
+
+protected theorem set.mem_centralizer_iff_subset (s : set S) {x : S} :
+  x ∈ s.centralizer ↔ s ⊆ centralizer x :=
+by simp only [set.centralizer, centralizer, set.mem_set_of_eq, @commute.symm_iff _ _ x]; refl
+
+protected theorem set.centralizer_eq (s : set S) :
+  s.centralizer =  ⋂ a ∈ s, centralizer a :=
+set.ext $ assume x,
+by simp only [s.mem_centralizer, set.mem_bInter_iff, mem_centralizer]
+
+protected theorem set.centralizer_decreasing {s t : set S} (h : s ⊆ t) :
+  t.centralizer ⊆ s.centralizer :=
+by simp only [set.centralizer_eq, set.bInter_subset_bInter_left h]
+
+end defs
+
+namespace commute
+
+section semigroup
+
+variables {S : Type*} [semigroup S]
+
+/-- If `a` commutes with both `b` and `c`, then it commutes with their product. -/
+protected theorem mul {a b c : S} (hab : commute a b) (hac : commute a c) : commute a (b * c) :=
+by dunfold commute; assoc_rw [hab.eq, hac.eq]
+
+/-- If both `a` and `b` commute with `c`, then their product commutes with `c`. -/
+theorem mul_left {a b c : S}
+  (hac : commute a c) (hbc : commute b c) : commute (a * b) c :=
+(hac.symm.mul hbc.symm).symm
+
+end semigroup
+
+protected theorem all {S : Type*} [comm_semigroup S] (a b : S) : commute a b := mul_comm a b
```

Comment:
Could you please provide an `example` of what should work once we have all these `@[simp]`?

Reply:
I mean, it would be nice to have
```lean
example {S : Type*} [semigroup S] {a b : S} (h : commute a b) : commute (a^5 * b^10) (a^7 * b^13) := by simp
```
but it doesn't work.

Reply:
Did you try `simp *`? ` I think you need `simp *` for `simp` lemmas with assumptions to trigger. Not sure if this will work though. 

 This lemma has no assumptions so it should always rewrite `commute a b` to `true` in a `comm_semigroup`. Again, I'm also not certain this will work.

Reply:
Works with `simp [h, h.symm]`. Probably there should be a tactic that does `simp only [h₁, h₁.symm, ...] with commute`, where `h₁, ...` are hypotheses of the form `commute _ _` but I'd postpone it till someone actually needs to prove that two long expressions commute.

Reply:
I added many `@[simp]` attributes. Could you please have a look at them?